vessel-runtime/src/main/java/com/textnow/android/vessel/VesselImpl.kt

```diff
@@ -133,46 +131,60 @@ class VesselImpl(
      *
      * @return true only if the entire database could be loaded into cache in the allotted timeout
      */
-    private fun preloadImpl(timeoutMS: Int?): Boolean {
+    private fun preloadImpl(timeoutMS: Int?): PreloadReport {
         val cursor = db.query("SELECT * FROM vessel", emptyArray())
 
         cursor.use {
             val typeCol = cursor.getColumnIndex("type")
             val dataCol = cursor.getColumnIndex("data")
 
             val startTimeMS = System.currentTimeMillis()
+            val report = PreloadReport()
+
             while (cursor.moveToNext()) {
                 val type = cursor.getString(typeCol)
                 val data = cursor.getString(dataCol)
 
                 val kclass = try {
                     Class.forName(type).kotlin
                 } catch (error: Exception) {
+                    report.missingTypes.add(type)
+                    report.errorsOcurred = true
+                    profiler.countBlocking(Event.TYPE_NOT_FOUND)
+                    continue
+                }
+
+                val obj = try {
+                    if (data != null) fromJson(data, kclass) else nullValue
+                } catch (error: Exception) {
+                    report.deserializationErrors.add(type)
+                    report.errorsOcurred = true
                     continue
                 }
 
-                val obj = if (data != null) fromJson(data, kclass) else nullValue
                 cache?.set(type, obj as Any)
 
                 timeoutMS?.let {
                     val duration = System.currentTimeMillis() - startTimeMS
                     if (duration > it) {
+                        report.timedOut = true
+                        profiler.countBlocking(Event.PRELOAD_TIMEOUT)
                         cursor.close()
-                        return false
+                        return report
```

Comment:
if we change this to break instead, I think we can remove the cursor.close since cursor.use should cover that use case.

Reply:
Good point.
Given `.use()` always closes the cursor, I don't think it matters whether this is a return or a break.
I did remove the `.close()` though.